File: Problems/0744. (E) Find Smallest Letter Greater Than Target/Java/solution.java
class Solution {
  public char nextGreatestLetter(char[] letters, char target) {
    // binary search;

    int lo = 0;
    int hi = letters.length - 1;
    int mid;
    while (lo <= hi) {
      mid = (lo + hi) / 2;
      if (letters[mid] > target)
        hi = mid-1;
      else
        lo = mid+1;
    }
    return lo != letters.length ? letters[lo] : letters[0];
  }
}

File: Problems/0744. (E) Find Smallest Letter Greater Than Target/Python3/solution.py
class Solution:
  def nextGreatestLetter(self, letters: List[str], target: str) -> str:
    # binary search
    
    lo, hi = 0, len(letters)-1
    while lo <= hi:
      mid = (hi + lo) // 2
      if letters[mid] > target:
        hi = mid-1
      else:
        lo = mid+1

    return letters[lo] if lo != len(letters) else letters[0]

